File: 2024/elixir/day2.exs
defmodule Main do 

  def part1() do
    input = File.read!("../inputs/day2.txt")
    
    lines = String.split(input, "\n", trim: true)

    ans = 0

    ans = Enum.reduce(lines,ans,fn line,ans->
      
      nums = String.split(line, " ", trim: true)

      nums = Enum.map(nums,fn i -> String.to_integer(i) end)

      flag1 = Enum.all?(for i <- 1..length(nums)-1 do
        if Enum.at(nums,i) < Enum.at(nums,i-1) && Enum.at(nums,i-1)-Enum.at(nums,i) >=1 && Enum.at(nums,i-1)-Enum.at(nums,i) <= 3 do
          true
        else
          false
        end
      end)

      flag2 = Enum.all?(for i <- 1..length(nums)-1 do
        if Enum.at(nums,i) > Enum.at(nums,i-1) && Enum.at(nums,i)-Enum.at(nums,i-1) >=1 && Enum.at(nums,i)-Enum.at(nums,i-1) <= 3 do
          true
        else
          false
        end
      end)

      if flag1 || flag2 do
        ans+1
      else
        ans
      end

    end)

    IO.puts("The answer for part 1 is: #{ans}")

  end

  def part2() do

    input = File.read!("../inputs/day2.txt")
    
    lines = String.split(input, "\n", trim: true)

    ans = 0

    ans = Enum.reduce(lines,ans,fn line,ans->
      
      nums = String.split(line, " ", trim: true)

      nums = Enum.map(nums,fn i -> String.to_integer(i) end)

      safe = Enum.any?(for i <- 0..length(nums) do 

        nnums = if i != length(nums) do
          List.delete_at(nums,i)
        else
          nums
        end

        flag1 = Enum.all?(for i <- 1..length(nnums)-1 do
          if Enum.at(nnums,i) < Enum.at(nnums,i-1) && Enum.at(nnums,i-1)-Enum.at(nnums,i) >=1 && Enum.at(nnums,i-1)-Enum.at(nnums,i) <= 3 do
            true
          else
            false
          end
        end)

        flag2 = Enum.all?(for i <- 1..length(nnums)-1 do
          if Enum.at(nnums,i) > Enum.at(nnums,i-1) && Enum.at(nnums,i)-Enum.at(nnums,i-1) >=1 && Enum.at(nnums,i)-Enum.at(nnums,i-1) <= 3 do
            true
          else
            false
          end
        end)

        flag1 || flag2

      end)

      if safe do
        ans+1
      else
        ans
      end

    end)

    IO.puts("The answer for part 2 is: #{ans}")
    
  end

end

Main.part1()
Main.part2()

File: 2024/rust/src/bin/day2.rs
use std::fs;

fn main() -> Result<(), Box<dyn std::error::Error>> {

    let input = fs::read_to_string("../inputs/day2.txt")?;

    let mut ans :i64 = 0;

    for i in input.lines() {
        let nums : Vec<i64> = i.split(" ").map(|x| x.parse::<i64>().unwrap()).collect();

        let mut safe1 : bool = true;
        let mut safe2 : bool = true;

        for j in 1..nums.len() {
            if nums[j]<nums[j-1] && nums[j-1]-nums[j]>=1 && nums[j-1]-nums[j]<=3 {
            }
            else{
                safe1 = false;
            }
            if nums[j]>nums[j-1] && nums[j]-nums[j-1]>=1 && nums[j]-nums[j-1]<=3 {
            }
            else{
                safe2 = false;
            }
        }

        if safe1 || safe2 {
            ans+=1;
            continue;
        }
    }

    println!("The answer for part 1 is: {}", ans);

    let mut ans :i64 = 0;

    'outer : for i in input.lines() {

        let nums : Vec<i64> = i.split(" ").map(|x| x.parse::<i64>().unwrap()).collect();
        
        for z in 0..nums.len()+1{

            let mut safe1 : bool = true;
            let mut safe2 : bool = true;

            let mut num = nums.clone();

            if z!=nums.len() {
                num.remove(z);
            }

            for j in 1..num.len() {
                if num[j]<num[j-1] && num[j-1]-num[j]>=1 && num[j-1]-num[j]<=3 {
                }
                else{
                    safe1 = false;
                }

                if num[j]>num[j-1] && num[j]-num[j-1]>=1 && num[j]-num[j-1]<=3 {
                }
                else{
                    safe2 = false;
                }
            }

            if safe1 || safe2 {
                ans+=1;
                continue 'outer;
            }
        }
    }

    println!("The answer for part 2 is: {}", ans);

    Ok(())
}
